File: Programmers/줄 서는 방법.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

long long factorial[21];

void pre_calc() {
	factorial[0] = 1;
	for (int i = 1; i <= 20; i++)
		factorial[i] = factorial[i - 1] * i;
}

vector<int> solution(int n, long long k) {
	pre_calc();
	vector<bool> chk(n, false);
	vector<int> answer;
	int select = 0;
	while (answer.size() != n) {
		for (int i = 0; i < n; i++) {
			if (chk[i]) continue;
			long long num = factorial[n - 1 - select];
			if (k > num) {
				k -= num;
				continue;
			}
			select++;
			answer.push_back(i + 1);
			chk[i] = true;
			break;

		}
	}
	return answer;
}

File: Programmers/줄 서는 방법.java
class Solution {
    long[] factorial = new long[21];

    public void preCalc() {
        factorial[0] = 1;
        for (int i = 1; i <= 20; i++)
            factorial[i] = factorial[i - 1] * i;
    }

    public int[] solution(int n, long k) {
        preCalc();
        int[] answer = new int[n];
        boolean[] check = new boolean[n + 1];
        int select = 0;
        while (select != n) {
            for (int i = 1; i <= n; i++) {
                if (check[i]) continue;
                long num = factorial[n - 1 - select];
                if (k > num) {
                    k -= num;
                    continue;
                }
                check[i] = true;
                answer[select++] = i;
                break;
            }
        }
        return answer;
    }
}